File: day05/Day05.scala
import scala.io.Source.stdin

def step(a: Int, b: Int) = if (a < b) a + 1 else if (a > b) a - 1 else a

@main def main(idx: Int) =
    val criteria : (a: Array[Int]) => Boolean =
      if idx == 1 then {case Array(x1,y1,x2,y2) => x1 == x2 || y1 == y2}
                  else _ => true
    println {
      stdin.getLines()
           .map{_.split(" -> ").flatMap(_.split(",").map(_.strip.toInt))}
           .filter(criteria)
           .flatMap{case Array(x1,y1,x2,y2) => {
             var result: List[(Int, Int)] = List()
             var x = x1
             var y = y1

             while ! (x == x2 && y == y2) do
               result = (y, x) :: result
               x = step(x, x2)
               y = step(y, y2)
             result = (y, x) :: result

             result
           }}
           .to(Iterable).groupBy(t => t).map{_._2.size}.filter{_ > 1}
           .size
    }

File: day05/Day05.rs
use std::collections::HashMap;
use std::env;
use std::io;

fn main() {
    let nums: Vec<Vec<u32>> = io::stdin()
        .lines()
        .map(|s| s.unwrap())
        .map(|s| {
            s.split(" -> ")
                .flat_map(|s| s.split(","))
                .map(str::parse)
                .map(Result::unwrap)
                .collect()
        })
        .collect();

    let step = |a: u32, b: u32| {
        if a < b {
            a + 1
        } else if a > b {
            a - 1
        } else {
            a
        }
    };

    let criteria: fn(&&Vec<u32>) -> bool = if env::args().nth(1).unwrap() == "1" {
        // Part one, use only horizontal and vertical
        |&v: &&Vec<u32>| match v[..] {
            [x1, y1, x2, y2] => x1 == x2 || y1 == y2,
            _ => false,
        }
    } else {
        // Part two, use all
        |_| true
    };

    let mut board: HashMap<(u32, u32), u32> = HashMap::new();

    for v in nums.iter().filter(criteria) {
        match v[..] {
            [x1, y1, x2, y2] => {
                let mut x = x1;
                let mut y = y1;

                while !(x == x2 && y == y2) {
                    *board.entry((y, x)).or_insert(0) += 1;

                    x = step(x, x2);
                    y = step(y, y2);
                }

                *board.entry((y, x)).or_insert(0) += 1;
            }
            _ => (),
        }
    }

    let count = board.iter().map(|kv| kv.1).filter(|&&v| v > 1).count();

    println!("{}", count);
}
